opm/flowdiagnostics/DerivedQuantities.cpp

```diff
@@ -0,0 +1,239 @@
+    /// volumes injected).
+    ///
+    /// The sweep efficiency is analogue to 1D displacement using
+    /// the F-Phi curve as flux function.
+    Graph sweepEfficiency(const Graph& flowcap_storagecap_curve)
+    {
+        const auto& storagecap = flowcap_storagecap_curve.first;
+        const auto& flowcap = flowcap_storagecap_curve.second;
+        if (flowcap.size() != storagecap.size()) {
+            throw std::runtime_error("sweepEfficiency(): Inconsistent sizes in input graph.");
+        }
+
+        // Compute tD and Ev simultaneously,
+        // skipping identical Phi data points.
+        const int n = flowcap.size();
+        std::vector<double> Ev;
+        std::vector<double> tD;
+        tD.reserve(n);
+        Ev.reserve(n);
+        tD.push_back(0.0);
+        Ev.push_back(0.0);
+        for (int ii = 1; ii < n; ++ii) { // Note loop limits.
+            const double fd = flowcap[ii] - flowcap[ii-1];
+            const double sd = storagecap[ii] - storagecap[ii-1];
+            if (fd != 0.0) {
+                tD.push_back(sd/fd);
+                Ev.push_back(storagecap[ii] + (1.0 - flowcap[ii]) * tD.back());
+            }
+        }
+
+        return std::make_pair(tD, Ev);
+    }
+
+
+
+
+
+    /// Compute pore volume associated with an injector-producer pair.
+    double injectorProducerPairVolume(const Toolbox::Forward& injector_solution,
+                                      const Toolbox::Reverse& producer_solution,
+                                      const std::vector<double>& pore_volume,
+                                      const CellSetID& injector,
+                                      const CellSetID& producer)
+    {
+        const auto& inj_tracer = injector_solution.fd.concentration(injector);
+        const auto& prod_tracer = producer_solution.fd.concentration(producer);
+
+        // Create sorted vector of injection tracer.
+        // A better CellSetValues could eliminate the need for this
+        // and the block below (for production tracer).
+        const int n_i = inj_tracer.cellValueCount();
+        std::vector<std::pair<int, double>> sorted_inj_vals(n_i);
+        for (int ii = 0; ii < n_i; ++ii) {
+            sorted_inj_vals[ii] = inj_tracer.cellValue(ii);
+        }
+        std::sort(sorted_inj_vals.begin(), sorted_inj_vals.end());
+
+        // Create sorted vector of production tracer.
+        const int n_p = prod_tracer.cellValueCount();
+        std::vector<std::pair<int, double>> sorted_prod_vals(n_p);
+        for (int ii = 0; ii < n_p; ++ii) {
+            sorted_prod_vals[ii] = prod_tracer.cellValue(ii);
+        }
+        std::sort(sorted_prod_vals.begin(), sorted_prod_vals.end());
+
+        int inj_pos = 0;
+        double volume = 0.0;
+        for (const auto cv : sorted_prod_vals) {
+            while (inj_pos < n_i && sorted_inj_vals[inj_pos].first < cv.first) {
+                ++inj_pos;
+            }
```

Comment:
I believe that might actually not be any faster for ranges normally encountered, but profiling would have to guide us there.

In any case, this code is scheduled to be replaced in #25 so I suggest we leave it unchanged for now.